### DISCUSSION TITLE
Using precomputed tSNE coordinates

### DISCUSSION THREAD

**Discussion by AAA-3** [theislab/scvelo#648] [Q&A] [answered]
Hello! I have seurat tSNE coordinates for my cells:

CellID | tSNE_1 | tSNE_2
-- | -- | --
221929_WT3:AACCATGCAGCCTTGGx | -23.0987642749178 | -15.9412999105341
221929_WT3:AACTGGTAGCGTAATAx | 15.4933217201963 | -18.9078173370809
221929_WT3:ACCCACTCAGGGTACAx | 9.28644825194571 | -21.3096658306121
221929_WT3:ACCTTTATCAATAAGGx | 4.2676676477722 | -21.5265299052195
221929_WT3:ACGAGGATCTACTTACx | -21.4905698959193 | -16.0147242794178
221929_WT3:ACTGCTCAGAAGATTCx | 2.33875227628558 | -22.0286597871625

And I want them to be transfered into my loom files which I concatenated:
![image](https://user-images.githubusercontent.com/68068335/129913175-c4b2ad0b-13a8-4619-bfdf-f3d60af4a32a.png)

How can I transfer my tsne coordinates onto the matching cell IDs in my concatenated loom file? I had a look at #37 but running the suggested code I got this really strange output... not entirely sure what `initial_size_spliced` `initial_size_unspliced` `initial_size` are or where they came from: 
![image](https://user-images.githubusercontent.com/68068335/129913409-bfbb3508-68a7-45ef-8fd8-8b4c8f892fce.png)

This is the first time I have tried using python so please let me know I haven't missed out on something!

**Comment by WeilerP** [comment]
@AAA-3, you should **not** use `scv.utils.merge` here. In your case, this will simply subset `sample_one` to observations present in both `sample_one` and `tsne`.
The columns `"initial_size"`, `"initial_size_unspliced"` and `"initial_size_spliced"` are added when calling `scvelo.utils.merge`. These are the counts per cell prior to subsetting, i.e. the initial size of the cell.

I'd do something along the lines of

```python
import pandas as pd

from anndata import read_loom

import scvelo as scv

WT3 = read_loom(...)
WT4 = read_loom(...)
# read rest of data, clean var names

sample_one = WT3.concatenate(...)

tsne_coords = pd.read_csv(...)

# Subset to observations present in both `sample_one` and `tsne_coords`
tsne_coords = tsne_coords.loc[tsne_coords.index.isin(sample_one.obs_names), :]
sample_one = sample_one[sample_one.obs_names.isin(tsne_coords.index), :]

# Make sure `sample_one` and `tsne_coords` are aligned
tsne_coords = tsne_coords.loc[sample_one.obs_names, :]

# Add tSNE coordinates to `sample_one`
sample_one.obsm['X_tsne'] = tsne_coords.values
```

Some thoughts/remarks:

1. You have `104414` observations in `sample_one` but tSNE coordinates for only `1611` of them. This strikes me as rather odd. Was this intended?
2. Of these observations, only `723` seem to appear in both objects (that's the size of the merged object). I am not sure but the concatenation step might tamper with the observation names. Maybe it would be better to first read the data, subset to observations present in the data and tSNE coordinates and then concatenate, i.e. something along the lines of

```python
import pandas as pd

from anndata import read_loom

import scvelo as scv

tsne_coords = pd.read_csv(...)
WT3 = read_loom(...)
WT3.var_names_make_unique()
tsne_coords = tsne_coords.loc[tsne_coords.index.isin(WT3.obs_names), :]
WT3 = WT3[WT3.obs_names.isin(tsne_coords.index), :]
WT3.obsm['X_tsne'] = tsne_coords.values

tsne_coords = pd.read_csv(...)
WT4 = read_loom(...)
...

sample_one = WT3.concatenate(...)
```

**Reply by AAA-3** [reply]
Hello! 

Thanks for you help! I did notice those issues you listed as well and here is the reason:

1. The discrepency is because in Seurat, we had really harsh quality controls which meant a lot of cells were not included in analysis AND also because Seurat included samples that we did not run Velocyto on.
2. The 723 observations was something I cannot explain...I am using Jupyter notebook and was using a tutorial (listed in my code below) to obtain those cell IDs. sometimes it would return with 43 (in some cases even 0) cells. Running the code again somehow returned all 1611 (see below):

Using your suggestion, I went in to subset the cells that were in my tSNE and CellID lists and the output for that was 0. I went back into my seurat output and manually sorted the cells by ID again and to try my luck I restarted the kernal and re-ran some old code which I was not having any luck with (adding below for reference):

```
#Following the tutorial: https://github.com/basilkhuder/Seurat-to-RNA-Velocity 
import anndata
import scvelo as scv
import pandas as pd
import numpy as np
import matplotlib as plt
import rpy2

#loading data
WT3 = anndata.read_loom("/home/ali/Dokumente/RPractise/E18.5_rawdata/E18.5_raw_outputs/221929_WT3/velocyto/221929_WT3.loom")
WT4 = anndata.read_loom("/home/ali/Dokumente/RPractise/E18.5_rawdata/E18.5_raw_outputs/222863_WT4/velocyto/222863_WT4.loom")
KO4 = anndata.read_loom("/home/ali/Dokumente/RPractise/E18.5_rawdata/E18.5_raw_outputs/222862_KO4/velocyto/222862_KO4.loom")
KO5 = anndata.read_loom("/home/ali/Dokumente/RPractise/E18.5_rawdata/E18.5_raw_outputs/222864_KO5/velocyto/222864_KO5.loom")
WT3.var_names_make_unique()
WT4.var_names_make_unique()
KO4.var_names_make_unique()
KO5.var_names_make_unique()

cellID_obs = pd.read_csv("/home/ali/Dokumente/RPractise/E18.5/MTW Expected E18.5/cellID_obs2.csv") #x from seurat; y&z manually sorted to look like WT3.obs.index
tsne = pd.read_csv("/home/ali/Dokumente/RPractise/E18.5/MTW Expected E18.5/cell_embeddings3.csv") #renaming tsne to umap for coding only
cell_clusters = pd.read_csv("/home/ali/Dokumente/RPractise/E18.5/MTW Expected E18.5/clusters.csv")

cellID_obs_WT3 = cellID_obs[cellID_obs["z"].str.contains("221929_WT3:")]
cellID_obs_WT4 = cellID_obs[cellID_obs["z"].str.contains("222863_WT4:")]
cellID_obs_KO4 = cellID_obs[cellID_obs["z"].str.contains("222862_KO4:")]
cellID_obs_KO5 = cellID_obs[cellID_obs["z"].str.contains("222864_KO5:")]
WT3 = WT3[np.isin(WT3.obs.index, cellID_obs_WT3["z"])]
WT4 = WT4[np.isin(WT4.obs.index, cellID_obs_WT4["z"])]
KO4 = KO4[np.isin(KO4.obs.index, cellID_obs_KO4["z"])]
KO5 = KO5[np.isin(KO5.obs.index, cellID_obs_KO5["z"])]
sample_one = WT3.concatenate(WT4, KO4, KO5,index_unique=None)

#cast our index as a data frame and change the column name
sample_one_index = pd.DataFrame(sample_one.obs.index)
sample_one = sample_one_index.rename(columns = {0:'CellID'})

#merge tsne data with concantenated
umap_combined = sample_one.merge(tsne, on = 'CellID')
#we're certain the orders are the same, we can remove the first column of data frame 
umap_combined = umap_combined.iloc[:,1:]

sample_one = WT3.concatenate(WT4, KO4, KO5,index_unique=None)
#and add the UMAP coordinates to our anndata object.
sample_one.obsm['X_tsne'] = umap_combined.values
sample_one.obsm['X_tsne'] 
```
successfully merging CellID and tSNE:
![image](https://user-images.githubusercontent.com/68068335/130077678-a5bed637-896a-4692-aa97-208b9d4cde8f.png)
Successfully bringing in the tSNE into the concatenated sample
![image](https://user-images.githubusercontent.com/68068335/130077792-fd74a5a8-8ee0-4c26-bbda-aa5067c33c7f.png)

Not sure why it worked this time round and not before, but might be that my manual sorting of CellID was to blame.
I am now going to try my hand at importing seurat clusters and hopefully I won't make the same mistake *fingers crossed*

**Comment by AAA-3** [comment]
Hello @WeilerP 

I was wondering if you could help me here since my issue is a continuation of the above. I managed to import my seurat clusters using the same methods as above.

However, when I try to load my saved loom file, I get an error each time I set the `obsm_mapping` parametre:

`adata = scv.read_loom("/home/ali/Dokumente/RPractise/Velocity/PythonCodes/adata.loom", obs_names="CellID", var_names="Gene", obsm_mapping="X_tsne")`
![image](https://user-images.githubusercontent.com/68068335/130209166-3b1126fd-529a-4026-a36a-1b3f3f149562.png)

Without setting the `obsm_mapping` parametre, the file loads normally and one can even see the tsne coordinates:
![image](https://user-images.githubusercontent.com/68068335/130209483-da0b29dc-82d9-4d16-ac4e-34250417f568.png)

Seeing as the information is available in the loom file, I assume that it is in an incorrect format...would it have something to do with it being an array? I notice that there is no `["tsne_1", "tsne_2"]`  headers, as listed in the [scvelo.read_loom API page](https://scvelo.readthedocs.io/scvelo.read_loom/)

**Reply by WeilerP** [reply]
@AAA-3, this is a Scanpy/anndata issue. Please open an issue/discussion there. `scvelo.read_loom` is a Scanpy function simply made available for convenience.

**Comment by AAA-3** [comment, marked as answer]
For anyone else who is having this issue: The problem was resolved when instead of saving it as a loom I saved it as a H5AD file :) Issue can be marked as solved
